contracts/golos.vesting/golos.vesting.cpp

```diff
@@ -0,0 +1,441 @@
+                account.modify(balance, 0, [&](auto &item){
+                    item.vesting += obj->amount;
+                });
+            } else {
+                account.emplace(_self, [&](auto &item){
+                    item.vesting = obj->amount;
+                });
+            }
+
+            obj = index.erase(obj);
+        } else {
+            break;
+        }
+    }
+}
+
+void vesting::create_pair(asset token, asset vesting) {
+    require_auth(_self);
+
+    auto index_token = _table_pair.get_index<N(token)>();
+    auto it_token = index_token.find(token.symbol.name());
+    eosio_assert(it_token == index_token.end(), "Pair with such a token already exists");
+
+    auto index = _table_pair.get_index<N(vesting)>();
+    auto it_vesting = index.find(vesting.symbol.name());
+    eosio_assert(it_vesting == index.end(), "Pair with such a vesting already exists");
+
+    _table_pair.emplace(_self, [&](auto &item){
+        item.id = _table_pair.available_primary_key();
+        item.vesting.symbol = vesting.symbol;
+        item.token.symbol   = token.symbol;
+    });
+}
+
+void vesting::issue(const structures::issue_vesting &m_issue, bool is_autorization, bool is_buy) {
+    if (is_autorization)
+        require_auth(_self);
+
+    auto sym = m_issue.quantity.symbol;
+    eosio_assert( sym.is_valid(), "invalid symbol name" );
+
+    auto sym_name = sym.name();
+    auto index = _table_pair.get_index<N(vesting)>();
+    auto pair = index.find(sym_name);
+    eosio_assert(pair != index.end(), "Vesting not found");
+
+    index.modify(pair, _self, [&](auto &item) {
+        item.vesting += m_issue.quantity;
+        if (is_buy)
+            item.token = convert_token(m_issue.quantity, pair->token.symbol);
+    });
+
+    add_balance(m_issue.to, m_issue.quantity, _self);
+}
+
+void vesting::transfer(account_name from, account_name to, asset quantity, bool is_autorization) {
+    eosio_assert(from != to, "cannot transfer to self");
+    if (is_autorization)
+        require_auth(from);
+    eosio_assert(is_account(to), "to account does not exist");
+
+    require_recipient(from);
+    require_recipient(to);
+
+    eosio_assert(quantity.is_valid(), "invalid quantity");
+    eosio_assert(quantity.amount > 0, "must transfer positive quantity");
+
+    auto sym_name = quantity.symbol.name();
+    auto index = _table_pair.get_index<N(token)>();
+    auto pair = index.find(sym_name);
+    eosio_assert(pair != index.end(), "Token not found");
+
+    sub_balance(from, convert_token(quantity, pair->vesting.symbol));
+    add_balance(to, convert_token(quantity, pair->vesting.symbol), from);
+}
+
+void vesting::sub_balance(account_name owner, asset value) {
+    tables::account_table account(_self, owner);
+
+    const auto& from = account.get(value.symbol.name(), "no balance object found");
+    eosio_assert(from.vesting >= value, "overdrawn balance");
+
+    if(from.vesting == value) {
+       account.erase(from);
```

Comment:
We can't erase info about account `from` if `vesting` is equal zero. User can have received_vesting or delegate_vesting.